manager/volume.go

```diff
@@ -229,73 +229,114 @@ func (m *VolumeManager) Attach(name, nodeID string, disableFrontend bool, attach
+	//	}
+	//	if v.Spec.EngineImage != v.Status.CurrentImage {
+	//		return nil, fmt.Errorf("upgrading in process for volume, cannot start migration")
+	//	}
+	//	if v.Spec.Standby || v.Status.IsStandby {
+	//		return nil, fmt.Errorf("dr volume migration is not supported")
+	//	}
+	//	if v.Status.ExpansionRequired {
+	//		return nil, fmt.Errorf("cannot migrate volume while an expansion is required")
+	//	}
+	//
+	//	v.Spec.MigrationNodeID = input.HostID
+	//	logrus.Infof("Volume %v migration from %v to %v requested", v.Name, v.Spec.NodeID, input.HostID)
+	//}
+	//
+	//v.Spec.DisableFrontend = input.DisableFrontend
+	//v.Spec.LastAttachedBy = input.AttachedBy
+	//v, err = m.ds.UpdateVolume(v)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	// TODO: special case for attachment by UI
+
+	va, err := m.ds.GetLHVolumeAttachment(types.GetLHVolumeAttachmentNameFromVolumeName(v.Name))
+	if err != nil {
+		return nil, err
 	}
 
-	isVolumeShared := v.Spec.AccessMode == longhorn.AccessModeReadWriteMany && !v.Spec.Migratable
-	isVolumeDetached := v.Spec.NodeID == ""
-	if isVolumeDetached {
-		if !isVolumeShared || disableFrontend {
-			v.Spec.NodeID = nodeID
-			logrus.Infof("Volume %v attachment to %v with disableFrontend %v requested", v.Name, v.Spec.NodeID, disableFrontend)
-		}
-	} else if isVolumeShared {
-		// shared volumes only need to be attached if maintenance mode is requested
-		// otherwise we just set the disabled frontend and last attached by states
-		logrus.Debugf("No need to attach volume %v since it's shared via %v", v.Name, v.Status.ShareEndpoint)
-	} else {
-		// non shared volume that is already attached needs to be migratable
-		// to be able to attach to a new node, without detaching from the previous node
-		if !v.Spec.Migratable {
-			return nil, fmt.Errorf("non migratable volume %v cannot attach to node %v is already attached to node %v", v.Name, nodeID, v.Spec.NodeID)
-		}
-		if v.Spec.MigrationNodeID != "" && v.Spec.MigrationNodeID != nodeID {
-			return nil, fmt.Errorf("unable to migrate volume %v from %v to %v since it's already migrating to %v",
-				v.Name, v.Spec.NodeID, nodeID, v.Spec.MigrationNodeID)
-		}
-		if v.Status.State != longhorn.VolumeStateAttached {
-			return nil, fmt.Errorf("invalid volume state to start migration %v", v.Status.State)
-		}
-		if v.Status.Robustness != longhorn.VolumeRobustnessHealthy && v.Status.Robustness != longhorn.VolumeRobustnessDegraded {
-			return nil, fmt.Errorf("volume must be healthy or degraded to start migration")
-		}
-		if v.Spec.EngineImage != v.Status.CurrentImage {
-			return nil, fmt.Errorf("upgrading in process for volume, cannot start migration")
-		}
-		if v.Spec.Standby || v.Status.IsStandby {
-			return nil, fmt.Errorf("dr volume migration is not supported")
-		}
-		if v.Status.ExpansionRequired {
-			return nil, fmt.Errorf("cannot migrate volume while an expansion is required")
-		}
+	logrus.Infof("=======================> 2")
```

Comment:
Remove this debugging comment. Also, other places.

Reply:
Thank you. Fixed